### PR TITLE
SunImagePlugin fixes

### LINKED ISSUE #2231: Sun Raster files with RLE encoding and depth 1 fail to load
### What did you do?
Tried to open a 1-bit RLE encoded Sun Raster file
### What did you expect to happen?
Expected image to load
### What actually happened?
Got a 'buffer overrun when reading image file' error
### What versions of Pillow and Python are you using?
Python 2.7.8, Pillow 3.3.1

Please include code that reproduces the issue and whenever possible, an image that demonstrates the issue. The best reproductions are self-contained scripts with minimal dependencies. If you are using a framework such as plone, django, or buildout, try to replicate the issue just using Pillow. 

Use image in attached ZIP file or create a new image in GIMP, 640x400 with white background. Select black brush, draw a line, convert Image->Mode to Indexed, and select black and white (1-bit) palette. Save image as Sun Raster file with RLE encoding.

[sunraster.zip](https://github.com/python-pillow/Pillow/files/597607/sunraster.zip)

```
>>> from PIL import Image
>>> im = Image.open('sunraster.im1')
>>> print im.format,im.size, im.mode
SUN (640, 400) 1
>>> im.load()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/rh/python27/root/usr/lib/python2.7/site-packages/PIL/ImageFile.py", line 240, in load
    raise_ioerror(err_code)
  File "/opt/rh/python27/root/usr/lib/python2.7/site-packages/PIL/ImageFile.py", line 59, in raise_ioerror
    raise IOError(message + " when reading image file")
IOError: buffer overrun when reading image file
>>>

```

### PR DESCRIPTION
**Pull request by wiredfool**
Fixes #2231 

Changes proposed in this pull request:

 * SunRle was borked, it couldn't have worked in the first place.
 * ImageFile.open would attempt to pass a too short args tuple into mmap. 
 * Ran a sample of 10 other images through the decoder, verified visually. Pre-fix, 2 worked. All work now. 

Tests.  I'd like to include the set of test images from https://samples.libav.org/image-samples/sunrast/ so that we don't regress. Libav is ok with mirroring/distribution, though half of the images are lena, and the others are of unclear license from upstream. So at least debian would have issues with them.

### CODE REVIEW

**Comment by hugovk** [comment]
If we can't include the test images in the source, one option is to zip them up onto https://github.com/python-pillow/pillow-depends and fetch them as part of the CI builds (and skip relevant tests if not present).

**Comment by wiredfool** [comment]
Ok, tests should be good, I've rebased away the two extra xcopy flag commits, so I'm done with this one.

File: Tests/test_file_sun.py [hunk 1]
```diff
@@ -16,10 +19,33 @@ def test_sanity(self):
...
+
+
+    @unittest.skipIf(not os.path.exists(EXTRA_DIR),
+                     "Extra image files not installed")
+    def test_others(self):
+        files = (os.path.join(EXTRA_DIR, f) for f in
+                   os.listdir(EXTRA_DIR) if os.path.splitext(f)[1]
+                   in ('.sun', '.SUN', '.ras'))
+        for path in files:
+            print (path)
```
Comment (hugovk) [inline Tests/test_file_sun.py]: Is this print needed?

File: PIL/SunImagePlugin.py [hunk 2]
```diff
@@ -48,31 +63,73 @@ def _open(self):
...
-
-        if i32(s[24:28]) != 0:
-            length = i32(s[28:32])
-            offset = offset + length
-            self.palette = ImagePalette.raw("RGB;L", self.fp.read(length))
+            raise SyntaxError("Unsupported Mode/Bit Depth")
+
+    
+        
+        if palette_length:
```
Comment (hugovk) [inline PIL/SunImagePlugin.py]: Can delete some of these blank lines.

File: PIL/SunImagePlugin.py [hunk 3]
```diff
@@ -48,31 +63,73 @@ def _open(self):
...
             self.mode = rawmode = "L"
         elif depth == 24:
-            self.mode, rawmode = "RGB", "BGR"
+            if file_type == 3:
+                self.mode, rawmode = "RGB", "RGB"
+            else:
+                self.mode, rawmode = "RGB", "BGR"
+        elif depth == 32:
+            if file_type == 3:
+                self.mode, rawmode = 'RGB', 'RGBX'
```
Comment (hugovk) [inline PIL/SunImagePlugin.py]: No tests for this line.

Comment (hugovk) [inline PIL/SunImagePlugin.py]: No tests for this line.

File: PIL/SunImagePlugin.py [hunk 4]
```diff
@@ -48,31 +63,73 @@ def _open(self):
...
         else:
-            raise SyntaxError("unsupported mode")
-
-        compression = i32(s[20:24])
-
-        if i32(s[24:28]) != 0:
-            length = i32(s[28:32])
-            offset = offset + length
-            self.palette = ImagePalette.raw("RGB;L", self.fp.read(length))
+            raise SyntaxError("Unsupported Mode/Bit Depth")
```
Comment (hugovk) [inline PIL/SunImagePlugin.py]: No tests for this line.

File: PIL/SunImagePlugin.py [hunk 5]
```diff
@@ -48,31 +63,73 @@ def _open(self):
...
+
+    
+        
+        if palette_length:
+            if palette_length > 1024:
+                raise SyntaxError("Unsupported Color Palette Length")
```
Comment (hugovk) [inline PIL/SunImagePlugin.py]: No tests for this line.

File: PIL/SunImagePlugin.py [hunk 6]
```diff
@@ -48,31 +63,73 @@ def _open(self):
...
+
+            if palette_type != 1:
+                raise SyntaxError("Unsupported Palette Type")
```
Comment (hugovk) [inline PIL/SunImagePlugin.py]: No tests for this line.

File: PIL/SunImagePlugin.py [hunk 7]
```diff
@@ -48,31 +63,73 @@ def _open(self):
...
+        # (http://www.fileformat.info/format/sunraster/egff.htm)
+
+        if file_type in (0, 1, 3, 4, 5):
             self.tile = [("raw", (0, 0)+self.size, offset, (rawmode, stride))]
-        elif compression == 2:
+        elif file_type == 2:
             self.tile = [("sun_rle", (0, 0)+self.size, offset, rawmode)]
-
+        else:
+            raise SyntaxError('Unsupported Sun Raster file type')
```
Comment (hugovk) [inline PIL/SunImagePlugin.py]: No tests for this line.

File: PIL/SunImagePlugin.py [hunk 8]
```diff
@@ -48,31 +63,73 @@ def _open(self):
...
-        compression = i32(s[20:24])
-
-        if i32(s[24:28]) != 0:
-            length = i32(s[28:32])
-            offset = offset + length
-            self.palette = ImagePalette.raw("RGB;L", self.fp.read(length))
+            raise SyntaxError("Unsupported Mode/Bit Depth")
+
+    
+
```
Comment (hugovk) [inline PIL/SunImagePlugin.py]: Can delete some of these blank lines.

File: PIL/SunImagePlugin.py (hunk 6 above)
Comment (wiredfool) [inline PIL/SunImagePlugin.py]: The palette bits are simple defensiveness.  We're doing an unprotected read here, not even a saferead.

**Pull request merged**